pandas/core/indexes/base.py

```diff
@@ -3569,6 +3569,7 @@ def _join_multi(self, other, how, return_indexers=True):
             # common levels, ldrop_names, rdrop_names
             dropped_names = ldrop_names + rdrop_names
 
+            assert isinstance(self, MultiIndex)  # for mypy
```

Comment:
could move up to just below `if self_is_mi ...` or do away with `self_is_mi` and `other_is_mi`